{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyzing Feb 2019 16s Sequencing Data\n",
    "## Written by Kathryn Cogert, Feb 5th, 2019\n",
    "\n",
    "### Background:\n",
    "USEARCH was used to clean up this sequencing data collected from several samples.  Namely-- \n",
    "\n",
    "Mini column samples:\n",
    " - KC AMX D46 (Day 46 granules from large anammox column reactor, taken just before mini column inoculation)\n",
    " - KC AC Minicolumn Upper (Day 23 granules taken from minicolumn. Red granules stripped of AC only)\n",
    " - KC AC Minicolumn Mixed (Day 23 granules taken from minicolumn. mixture of black and red granules)\n",
    " - KC AC Batch (Granules from the failed batch experiment)\n",
    "\n",
    "Red & Grey Anammox Granules:\n",
    " - KC AMX D46 (Day 46 granules from large anammox column reactor, taken just before mini column inoculation)\n",
    " - KC Red (Red granules taken from large anammox column reactor)\n",
    " - KC Grey (Grey granules taken from large anammox column reactor)\n",
    "\n",
    "### Objective:\n",
    "This analysis is intended to drop sequencing results into a CSV for data visualization in R. Preprocessing of data was done with USEARCH - see in [Pipeline.ipynb](Pipeline.ipynb))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from ipywidgets import interact, interactive, fixed, interact_manual\n",
    "import numpy as np\n",
    "import ipywidgets as widgets\n",
    "%matplotlib inline\n",
    "plt.rcParams['figure.figsize'] = (10,7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 1) Load in OTU counts per sample table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OTU #</th>\n",
       "      <th>KC.AC.MiniColumn.Upper</th>\n",
       "      <th>KC.AMX.D46</th>\n",
       "      <th>KC.AMX.Grey</th>\n",
       "      <th>KC.AMX.Red</th>\n",
       "      <th>KC.AC.Batch</th>\n",
       "      <th>KC.AC.MiniColumn.Mixed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Otu1</td>\n",
       "      <td>4547</td>\n",
       "      <td>6856</td>\n",
       "      <td>1276</td>\n",
       "      <td>8603</td>\n",
       "      <td>3822</td>\n",
       "      <td>3209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Otu34</td>\n",
       "      <td>312</td>\n",
       "      <td>229</td>\n",
       "      <td>364</td>\n",
       "      <td>119</td>\n",
       "      <td>10</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Otu18</td>\n",
       "      <td>797</td>\n",
       "      <td>494</td>\n",
       "      <td>374</td>\n",
       "      <td>296</td>\n",
       "      <td>903</td>\n",
       "      <td>624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Otu22</td>\n",
       "      <td>634</td>\n",
       "      <td>486</td>\n",
       "      <td>407</td>\n",
       "      <td>203</td>\n",
       "      <td>320</td>\n",
       "      <td>535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Otu331</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    OTU #  KC.AC.MiniColumn.Upper  KC.AMX.D46  KC.AMX.Grey  KC.AMX.Red  \\\n",
       "0    Otu1                    4547        6856         1276        8603   \n",
       "1   Otu34                     312         229          364         119   \n",
       "2   Otu18                     797         494          374         296   \n",
       "3   Otu22                     634         486          407         203   \n",
       "4  Otu331                       9           2           10           1   \n",
       "\n",
       "   KC.AC.Batch  KC.AC.MiniColumn.Mixed  \n",
       "0         3822                    3209  \n",
       "1           10                     150  \n",
       "2          903                     624  \n",
       "3          320                     535  \n",
       "4            1                       1  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "otu_count = pd.read_csv('AMX.zotutab.txt', sep='\\t').rename(columns={'#OTU ID':'OTU #'})\n",
    "otu_count.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 2) Load in the OTU taxonomy assignment table & data clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OTU #</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Otu3</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Cytophagales</td>\n",
       "      <td>Microscillaceae</td>\n",
       "      <td>uncultured</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Otu2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Betaproteobacteriales</td>\n",
       "      <td>Rhodocyclaceae</td>\n",
       "      <td>Denitratisoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Otu1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Planctomycetes</td>\n",
       "      <td>Brocadiae</td>\n",
       "      <td>Brocadiales</td>\n",
       "      <td>Scalinduaceae</td>\n",
       "      <td>Candidatus_Scalindua</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Otu5</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Planctomycetes</td>\n",
       "      <td>Brocadiae</td>\n",
       "      <td>Brocadiales</td>\n",
       "      <td>Scalinduaceae</td>\n",
       "      <td>Candidatus_Scalindua</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Otu4</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Betaproteobacteriales</td>\n",
       "      <td>Rhodocyclaceae</td>\n",
       "      <td>Denitratisoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  OTU #    Domain          Phylum                Class                  Order  \\\n",
       "0  Otu3  Bacteria   Bacteroidetes          Bacteroidia           Cytophagales   \n",
       "1  Otu2  Bacteria  Proteobacteria  Gammaproteobacteria  Betaproteobacteriales   \n",
       "2  Otu1  Bacteria  Planctomycetes            Brocadiae            Brocadiales   \n",
       "3  Otu5  Bacteria  Planctomycetes            Brocadiae            Brocadiales   \n",
       "4  Otu4  Bacteria  Proteobacteria  Gammaproteobacteria  Betaproteobacteriales   \n",
       "\n",
       "            Family                 Genus  \n",
       "0  Microscillaceae            uncultured  \n",
       "1   Rhodocyclaceae         Denitratisoma  \n",
       "2    Scalinduaceae  Candidatus_Scalindua  \n",
       "3    Scalinduaceae  Candidatus_Scalindua  \n",
       "4   Rhodocyclaceae         Denitratisoma  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "otu_id = pd.read_table('AMX.otus.sintax', header=None, delimiter='\\t', index_col=None)\n",
    "otu_id_clean = otu_id. \\\n",
    "join(otu_id[3].str.split(',', 5, expand=True). \\\n",
    "rename(columns={0:'Domain',1:'Phylum',2:'Class',3:'Order',4:'Family',5:'Genus'})). \\\n",
    "drop(columns=[1,2,3]). \\\n",
    "rename(columns={0:'OTU #'})\n",
    "otu_id_clean.iloc[:,1:] = otu_id_clean.iloc[:,1:].apply(lambda x: x.str.replace('[a-z]:',''))\n",
    "\n",
    "otu_id_clean.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 3) Merge two dataframes together and normalize columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OTU #</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>KC.AC.MiniColumn.Upper</th>\n",
       "      <th>KC.AMX.D46</th>\n",
       "      <th>KC.AMX.Grey</th>\n",
       "      <th>KC.AMX.Red</th>\n",
       "      <th>KC.AC.Batch</th>\n",
       "      <th>KC.AC.MiniColumn.Mixed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Otu1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Planctomycetes</td>\n",
       "      <td>Brocadiae</td>\n",
       "      <td>Brocadiales</td>\n",
       "      <td>Scalinduaceae</td>\n",
       "      <td>Candidatus_Scalindua</td>\n",
       "      <td>0.097341</td>\n",
       "      <td>0.112858</td>\n",
       "      <td>0.027443</td>\n",
       "      <td>0.227852</td>\n",
       "      <td>0.090724</td>\n",
       "      <td>0.084726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Otu10</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Planctomycetes</td>\n",
       "      <td>Phycisphaerae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.035023</td>\n",
       "      <td>0.030535</td>\n",
       "      <td>0.007398</td>\n",
       "      <td>0.007045</td>\n",
       "      <td>0.046572</td>\n",
       "      <td>0.038812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Otu100</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Flavobacteriales</td>\n",
       "      <td>Cryomorphaceae</td>\n",
       "      <td>uncultured</td>\n",
       "      <td>0.001927</td>\n",
       "      <td>0.000741</td>\n",
       "      <td>0.000409</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Otu101</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Betaproteobacteriales</td>\n",
       "      <td>Nitrosomonadaceae</td>\n",
       "      <td>mle17</td>\n",
       "      <td>0.000257</td>\n",
       "      <td>0.000823</td>\n",
       "      <td>0.001204</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>0.001258</td>\n",
       "      <td>0.000370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>Otu102</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004178</td>\n",
       "      <td>0.000026</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      OTU #    Domain          Phylum                Class  \\\n",
       "2      Otu1  Bacteria  Planctomycetes            Brocadiae   \n",
       "9     Otu10  Bacteria  Planctomycetes        Phycisphaerae   \n",
       "99   Otu100  Bacteria   Bacteroidetes          Bacteroidia   \n",
       "100  Otu101  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "103  Otu102  Bacteria      Firmicutes           Clostridia   \n",
       "\n",
       "                     Order             Family                 Genus  \\\n",
       "2              Brocadiales      Scalinduaceae  Candidatus_Scalindua   \n",
       "9                     None               None                  None   \n",
       "99        Flavobacteriales     Cryomorphaceae            uncultured   \n",
       "100  Betaproteobacteriales  Nitrosomonadaceae                 mle17   \n",
       "103          Clostridiales               None                  None   \n",
       "\n",
       "     KC.AC.MiniColumn.Upper  KC.AMX.D46  KC.AMX.Grey  KC.AMX.Red  KC.AC.Batch  \\\n",
       "2                  0.097341    0.112858     0.027443    0.227852     0.090724   \n",
       "9                  0.035023    0.030535     0.007398    0.007045     0.046572   \n",
       "99                 0.001927    0.000741     0.000409    0.000026     0.000000   \n",
       "100                0.000257    0.000823     0.001204    0.000689     0.001258   \n",
       "103                0.000021    0.000000     0.000022    0.000000     0.004178   \n",
       "\n",
       "     KC.AC.MiniColumn.Mixed  \n",
       "2                  0.084726  \n",
       "9                  0.038812  \n",
       "99                 0.001663  \n",
       "100                0.000370  \n",
       "103                0.000026  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = otu_id_clean.merge(right=otu_count, on='OTU #')\n",
    "colsum = pd.DataFrame(df.sum(axis=0)[1:7]).transpose()\n",
    "df.iloc[:,7:] = df.iloc[:,7:] /  df.iloc[:,7:].sum()\n",
    "df.sort_values('OTU #').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 4) Note unidentified organisms as such"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.fillna('Unidentified', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 5) Save as CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('AC_Column_Abundances.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplemental\n",
    "\n",
    "Previously used interactive plot to visualize by taxonomic rank."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "003e8681e49a4a7eb52c270c909bb689",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='Rank', index=3, options=('Domain', 'Phylum', 'Class', 'Order', 'Fa…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@interact(Rank=['Domain', 'Phylum', 'Class', 'Order', 'Family', 'Genus'],\n",
    "         Experiment = ['Activated Carbon', 'Granule Color'])\n",
    "def compare_v(Rank='Order', Experiment='Activated Carbon'): \n",
    "    #to_plot = (df.groupby([Rank]).sum()/ df. \\\n",
    "    #           groupby([Rank]).sum().sum()). \\\n",
    "    #reindex(sorted(df.columns), axis=1)\n",
    "    to_plot = df.groupby([Rank]).sum().sort_values('KC.AMX.D46', \n",
    "                                                   ascending=False).transpose()\n",
    "    if Experiment=='Activated Carbon':\n",
    "        to_plot = to_plot.loc[['KC.AMX.D46',\n",
    "                               'KC.AC.MiniColumn.Upper',\n",
    "                               'KC.AC.MiniColumn.Mixed', \n",
    "                               'KC.AC.Batch']]\n",
    "    elif Experiment == 'Granule Color':\n",
    "        to_plot = to_plot.loc[['KC.AMX.D46',\n",
    "                               'KC.AMX.Grey', \n",
    "                               'KC.AMX.Red']]\n",
    "    fig=to_plot.plot(kind='bar', stacked=True)\n",
    "    \n",
    "    #for i in [0,2,4,6,8, 10]: \n",
    "    # Omit KC.AMXRT results b/c not in v10 results.\n",
    "    #    fig=to_plot.iloc[i:i+2,:].plot(kind='bar', stacked=True)\n",
    "    plt.title(Experiment, color='white', fontsize=14, fontweight='bold')\n",
    "    plt.xticks(color='white', size=16, rotation=45)\n",
    "    plt.yticks(np.arange(0,1.1,0.1), color='white')\n",
    "    fig.legend(bbox_to_anchor=(0.75, -0.1), ncol=2)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
